File: oi28/gan/gan.py
from collections import defaultdict, deque
from math import sqrt, floor
from typing import Tuple, List, Optional, Set


def read_ints():
    return list(map(int, input().split(" ")))


class State:
    def __init__(self, n: int):
        self.size = 0
        self._state = [0] * n

    def add(self, x: int):
        if self._state[x] == 0:
            self.size += 1
        self._state[x] += 1

    def delete(self, x: int):
        self._state[x] -= 1
        if self._state[x] == 0:
            self.size -= 1


counter = 0
tree = defaultdict(list)
dfs_order: List[Tuple[int, bool]] = []
preorder: List[Optional[int]]
parent: List[int]


def get_edge(current_node, next_node) -> Tuple[int, int]:
    return (next_node, current_node) if next_node < current_node else (current_node, next_node)


def dfs(start_node: int):  # we need iterative version as recursive version throws RecursionError :<
    global counter, dfs_order
    q = deque([start_node])
    while len(q) != 0:
        node = q.popleft()
        if preorder[node] == -1:  # not visited
            q.appendleft(node)  # add again to calculate exit time
            dfs_order[counter] = (node, True)  # enter node
            preorder[node] = counter  # store preorder
            counter += 1
            for neighbour in tree[node]:
                if preorder[neighbour] == -1:  # not visited yet
                    parent[neighbour] = node  # mark parent
                    q.appendleft(neighbour)
        else:
            dfs_order[counter] = (node, False)  # exit node
            counter += 1


def edge_in_path(edge: Tuple[int, int], path: Set[int]) -> bool:
    return edge[0] in path and edge[1] in path


def main():
    n, m, z = read_ints()
    state = State(m)
    edges = {}
    edges_content = {}
    # read tree structure
    for i in range(n - 1):
        edge_from, edge_to, mascot = read_ints()
        edge_from, edge_to = edge_from - 1, edge_to - 1  # decrease
        edge = get_edge(edge_from, edge_to)
        edges[i] = edge
        edges_content[edge] = mascot - 1  # decrease for zero-based indexing
        tree[edge_from].append(edge_to)
        tree[edge_to].append(edge_from)
    queries: List[Tuple[int, int]] = []
    changes: List[Tuple[int, int, int]] = []
    # read queries
    time = 0
    for i in range(z):
        query = input().split(" ")
        if query[0] == 'Z':  # query
            node = int(query[1]) - 1  # decrease for zero-based indexing
            queries.append((time, node))
        else:  # q[0] == 'B' change mascot on edge
            time += 1
            edge_id, next_mascot = map(lambda x: int(x) - 1, query[1:])  # decrease for zero-based indexing
            edge = edges[edge_id]
            current_mascot = edges_content[edge]
            changes.append((edge_id, current_mascot, next_mascot))
            edges_content[edge] = next_mascot

    # init arrays
    global preorder, parent, dfs_order
    preorder = [-1] * n
    parent = [0] * n
    parent[0] = 0
    dfs_order = [(-1, False)] * (2 * n)  # 0(2n) memory, we can afford it, but we need to preallocate it
    # assign node ids, determine parent
    dfs(0)

    # move through (t,v) space
    current_t = time  # current tree state is same as in finish time
    current_v = 0  # current dfs order, 1 root is already added
    current_node = 0  # current node, 0 because we start from root
    current_path = {0}  # current path, only 0 because we start from root

    def add_node_to_path(node_to_add: int):
        nonlocal current_node
        edge = get_edge(node_to_add, current_node)
        mascot_to_add = edges_content[edge]
        state.add(mascot_to_add)
        current_path.add(node_to_add)
        current_node = node_to_add

    def remove_last_node_from_path():
        nonlocal current_node
        edge = get_edge(current_node, parent[current_node])
        mascot_to_del = edges_content[edge]
        state.delete(mascot_to_del)
        current_path.remove(current_node)
        current_node = parent[current_node]

    # sort queries
    sqrt_n = sqrt(n)
    # as current tree state is in last time stamp, we will sort by -t/sqrt_n
    ordered_queries = map(lambda x: x[2], sorted((-floor(q_t / sqrt_n), q_v, q_idx)
                                                 for q_idx, (q_t, q_v) in enumerate(queries)))

    answers: List[int] = [0] * len(queries)  # init answers array as result will be calculated out of order
    for q_idx in ordered_queries:
        qt, qv = queries[q_idx]
        while current_t < qt:  # shift state to future time
            edge_id, prev_mascot, next_mascot = changes[current_t]  # change mascot prev -> next on edge
            edge = edges[edge_id]
            if edge_in_path(edge, current_path):  # update state if change is in current path
                state.delete(prev_mascot)
                state.add(next_mascot)
            edges_content[edge] = next_mascot
            current_t += 1
        while current_t > qt:  # shift state to past time
            edge_id, prev_mascot, next_mascot = changes[current_t - 1]  # change mascot next -> prev on edge
            edge = edges[edge_id]
            if edge_in_path(edge, current_path):  # update state if change is in current path
                state.delete(next_mascot)
                state.add(prev_mascot)
            edges_content[edge] = prev_mascot
            current_t -= 1
        while current_v < preorder[qv]:  # need to move dfs forward
            next_node, next_operation = dfs_order[current_v + 1]  # node id, added/removed
            if next_operation:  # go down the tree
                add_node_to_path(next_node)
            else:  # return to node's parent
                remove_last_node_from_path()
            current_v += 1
        while current_v > preorder[qv]:  # need to revert dfs
            last_node, last_operation = dfs_order[current_v]  # node id, added/removed
            if last_operation:  # last node was added, so to reverse it we have to remove it
                remove_last_node_from_path()
            else:
                add_node_to_path(last_node)
            current_v -= 1
        answers[q_idx] = state.size  # store number of distinct mascots in path
    for answer in answers:
        print(answer)


if __name__ == '__main__':
    main()

File: oi28/gan/gan.cpp
#include <vector>
#include <iostream>
#include <tuple>
#include <deque>
#include <set>
using namespace std;

typedef pair<int, bool> pib;
typedef pair<int, int> pii;

struct Query
{
    int time, nodeId;
};

struct Change
{
    int edgeId, previusMascot, nextMascot;
};

struct DfsMove
{
    int nodeId, enter;
};

vector<int> state;
int stateSize = 0;
const int MAX_N = 100000,
          MAX_M = 150000,
          MAX_Z = 150000;
vector<int> tree[MAX_N];
int edgeContent[MAX_N * 2];
vector<Query> queries;
vector<Change> changes;
int preorder[100000];
int parent[100000];
int preorderCounter = 0;
vector<DfsMove> dfsOrder;
vector<pii> edges;
set<int> visitedEdges;
set<int> visitedNodes;
vector<int> childNodeIdToEdgeId;

void stateAdd(int x)
{
    cout << "state add " << x << "\n";
    if (state[x] == 0)
    {
        stateSize++;
    }
    state[x]++;
}

void stateDel(int x)
{
    cout << "state del " << x << "\n";
    state[x]--;
    if (state[x] == 0)
    {
        stateSize--;
    }
}

void dfs(int startNode)
{
    cout << "startNode in dfs" << startNode << "\n";
    dfsOrder.push_back({startNode, true});
    for (int edgeId : tree[startNode])
    {
        pii edge = edges[edgeId];
        int neighbour = (edge.first == startNode) ? edge.second : edge.first;
        if (preorder[neighbour] == 0 && neighbour != 0)
        {
            preorder[neighbour] = ++preorderCounter;
            parent[neighbour] = startNode;
            dfs(neighbour);
        }
    }
    preorderCounter++;
    dfsOrder.push_back({startNode, false});
}

void calulateChildNodeIdToEdgeId()
{
    for (int i = 0; i < edges.size(); i++)
    {
        auto edge = edges[i];
        int child;
        if (parent[edge.first] == edge.second)
        {
            child = edge.first;
        }
        else
        {
            child = edge.second;
        }
        childNodeIdToEdgeId[child] = i;
    }
}

int addEdge(int nodeId)
{
    cout << "addEgde(" << nodeId <<")\n";
    int edgeId = childNodeIdToEdgeId[nodeId];
    cout << "edge Id " << edgeId << "\n";
    int mascot = edgeContent[edgeId];
    stateAdd(mascot);
    cout << "add mascot:" << mascot << "\n";
    visitedNodes.insert(nodeId);
    visitedEdges.insert(edgeId);
    cout << "new current node" << nodeId << "\n";
    return nodeId;
}

int delEdge(int lastNodeId)
{
    cout << "delEgde(" << lastNodeId <<")\n";
    int edgeId = childNodeIdToEdgeId[lastNodeId];
    cout << "edge Id " << edgeId << "\n";
    int mascot = edgeContent[edgeId];
    stateDel(mascot);
    cout << "del mascot:" << mascot << "\n";
    visitedNodes.erase(lastNodeId);
    visitedEdges.erase(edgeId);
    cout << "new current node " << parent[lastNodeId] << "\n";
    return parent[lastNodeId];
}

int main(int argc, char const *argv[])
{
    int n, m, z;
    cin >> n >> m >> z;
    state.reserve(m);
    queries.reserve(z);
    changes.reserve(z);
    edges.resize(n - 1);
    childNodeIdToEdgeId.resize(n - 1);
    for (int i = 0; i < n - 1; i++)
    {
        int nodeFrom, nodeTo, mascotId;
        cin >> nodeFrom >> nodeTo >> mascotId;
        nodeFrom--;
        nodeTo--;
        edgeContent[i] = mascotId;
        edges[i] = {nodeFrom, nodeTo};
        tree[nodeFrom].push_back(i); // store edge id instead of node id
        tree[nodeTo].push_back(i);
    }
    // process tree
    parent[0] = -1;
    dfs(0);
    childNodeIdToEdgeId[0] = -1;
    calulateChildNodeIdToEdgeId();
    cout << "parent ";
    for (int i = 0; i < n; i++)
    {
        cout << parent[i] << " ";
    }
    cout << "\n";
    cout << "preorder ";
    for (int i = 0; i < n; i++)
    {
        cout << preorder[i] << " ";
    }
    cout << "\n";

    cout << "dfs order ";
    for (int i = 0; i < dfsOrder.size(); i++)
    {
        cout << " (" << ((dfsOrder[i].enter) ? 't' : 'f') << " " << dfsOrder[i].nodeId << ")";
    }
    cout << "\n";

    // read queries
    int time = 0;
    for (int i = 0; i < z; i++)
    {
        char type;
        cin >> type;
        cout << "type " << type << "\n";
        switch (type)
        {
        case 'Z':
            int query;
            cin >> query;
            query--;
            queries.push_back({time, query});
            break;
        default: // 'B'
            int edgeId, nextMascot;
            cin >> edgeId >> nextMascot;
            edgeId--;
            int currentMascot = edgeContent[edgeId];
            changes.push_back({edgeId, currentMascot, nextMascot});
            edgeContent[edgeId] = nextMascot;
            time++;
            break;
        }
    }
    int currentTime = time;
    int currentDfsStage = 0; // start when root already added
    int currentNode = 0;
    vector<int> answers(queries.size());
    for (int i = 0; i < queries.size(); i++)
    {
        auto query = queries[i];
        cout << "q.time " << query.time << " q.nodeId " << query.nodeId << " currentTime:" << currentTime << " currentNode:" << currentNode << "\n";
        cout << "edgesContent ";
        for (int i = 0; i < n - 1; i++)
        {
            cout << edgeContent[i] << " ";
        }
        cout << "\n";

        cout << "child to edge id ";
        for (int i = 0; i < n; i++)
        {
            cout << childNodeIdToEdgeId[i] << " ";
        }
        cout << "\n";
        while (currentTime < query.time)
        { // forward in time
            auto change = changes[currentTime++];
            cout << "forward in time edge:" << change.edgeId << " " << change.previusMascot << " -> " << change.nextMascot << "\n";
            int mascot = edgeContent[change.edgeId];
            if (visitedEdges.find(change.edgeId) != visitedEdges.end())
            {
                stateDel(change.previusMascot);
                stateAdd(change.nextMascot);
            }
            edgeContent[change.edgeId] = change.nextMascot;
        }
        while (currentTime > query.time)
        { // back in time
            auto change = changes[--currentTime];
            cout << "back in time edge:" << change.edgeId << " " << change.previusMascot << " -> " << change.nextMascot << "\n";
            int mascot = edgeContent[change.edgeId];
            if (visitedEdges.find(change.edgeId) != visitedEdges.end())
            {
                stateDel(change.nextMascot);
                stateAdd(change.previusMascot);
            }
            edgeContent[change.edgeId] = change.previusMascot;
        }
        while (currentDfsStage < preorder[query.nodeId])
        {
            DfsMove nextDfsMove = dfsOrder[++currentDfsStage];
            cout << "dfs move " << ((nextDfsMove.enter) ? "true" : "false") << " " << nextDfsMove.nodeId << "\n";
            if (nextDfsMove.enter)
            { // add edge and its content
                currentNode = addEdge(nextDfsMove.nodeId);
            }
            else
            {
                currentNode = delEdge(currentNode);
            }
        }
        while (currentDfsStage > preorder[query.nodeId])
        {
            DfsMove nextDfsMove = dfsOrder[currentDfsStage--];
            cout << "reverse dfs move currentDfsStage:" << currentDfsStage << " " << ((nextDfsMove.enter) ? "true" : "false") << " " << nextDfsMove.nodeId << "\n";
            if (nextDfsMove.enter)
            { // remove edge and its content
                currentNode = delEdge(currentNode);
            }
            else
            {
                currentNode = addEdge(nextDfsMove.nodeId);
            }
        }
        cout << "stateSize" << stateSize << "\n";
        answers[i] = stateSize;
    }
    for (auto ans : answers)
    {
        cout << ans << "\n";
    }
    return 0;
}
